Search and Query blocks: Add support for Default queries via `pre_get_posts` filter

michalczaplinski opened the pull request:
## What?
This adds support for Default queries via `pre_get_posts` filter. 

## Why?
The PR implementing the Instant Search ([#67181](https://github.com/WordPress/gutenberg/pull/67181)) does not handle the Default (inherited) queries.

## How?

The Default queries use the `?instant-search=<search-term>` URL search param.

michalczaplinski committed:
Add support for multiple search queries via `pre_get_posts` filter

michalczaplinski committed:
Add the `is_admin()` checks

michalczaplinski committed:
Enhance search block functionality by integrating canonical URL support and updating interactivity configuration. This change allows the search block to utilize the canonical URL when performing instant searches.

michalczaplinski commented:
I've tried to get the canonical URL of the requested page by using `get_permalink()` in [f4a6644](https://github.com/WordPress/gutenberg/pull/67289/commits/f4a6644650f832dff8b3aa51f2e3babd080f7d0a)

This approach was discussed previously in https://github.com/WordPress/gutenberg/pull/63147#issuecomment-2510896654

Unfortunately, this approach [does not](https://developer.wordpress.org/reference/functions/get_permalink/#:~:text=Note%20that%20when%20used%20outside%20The%20Loop%20on%20a%20posts%20page%20(index%2C%20archive%2C%20etc.)%20without%20the%20ID%20parameter%2C%20it%20will%20return%20the%20URL%20of%20the%20last%20post%20in%20The%20Loop%2C%20not%20the%20permalink%20for%20the%20current%20page.) work well: 

> Note that when used outside The Loop on a posts page (index, archive, etc.) without the ID parameter, it will return the URL of the last post in The Loop, not the permalink for the current page.

luisherranz commented:
Can't you pass the post ID manually?

michalczaplinski commented:
There is no post ID when the user is on an archive page (home, categories archive, etc.). This is when `get_permalink()` does not work.

I don't currently have a good idea how to overcome this but I _think_ there should be a way to retrieve the "canonical" URL of the page for a given URL. I see that there's some related logic in `class-wp-rewrite.php` and in https://developer.wordpress.org/reference/functions/redirect_canonical/ but that doesn't quite do what we want either.

luisherranz commented:
> but I _think_ there should be a way to retrieve the "canonical" URL of the page for a given URL

Yeah, I'm sure there has to be a way to do so too 😄

Maybe we need different strategies for different types of content.

Inside terms, something more like this? (untested)

```php
$term = get_queried_object();
if ( isset( $term->term_id ) ) {
  $url = get_term_link( $term->term_id );
}
```

michalczaplinski commented:
ok, I think I got it to work using: 

```php
// Get the canonical URL without pagination, works for archive pages
$canonical_url_no_pagination = get_pagenum_link(1);

// If we're on a singular post/page, use its permalink instead
if (is_singular()) {
	$canonical_url_no_pagination = get_permalink();
}
```

michalczaplinski committed:
Handle pagination correctly by using the canonical URL without pagination.

michalczaplinski committed:
Add e2e tests for instant search with default query.

michalczaplinski commented:
Currently working on adding support for multiple Default (inherited) queries on a single page. 

We'll have to assign an ID to each query, even if they are inherited. The editor automatically adds a `queryId` **attribute** to each query (even inherited ones). So, it looks like we can just use that but we'd have to pass that ID from the block context to the global query (which might be a bit tricky).

I'm trying something like this:

```php
function gutenberg_block_core_query_add_query_id_to_context( $context, $block ) {
	global $wp_query;

	if ( $block['blockName'] === 'core/query' && isset( $context['queryId'] ) ) {
		$wp_query->query_vars['query_id'] = $context['queryId'];
	}
	return $context;
}
add_filter( 'render_block_context', 'gutenberg_block_core_query_add_query_id_to_context', 10, 2 );
``` 

And then read that `query_id` in the `pre_get_posts` filter.


```php
function gutenberg_block_core_query_add_search_query_filtering( $query ) {
        // ...
	$query_id = $query->query_vars['query_id'];
	// ...
}
add_action(
	'pre_get_posts',
	'gutenberg_block_core_query_add_search_query_filtering'
);
```


but I have some doubts as to whether it's a good idea to create side-effects like this inside of a function passed to `add_filter()`. Filters should ideally be pure functions.

michalczaplinski commented:
Giving it further thought, I _think_ we don't have to support multiple Instant Search blocks by giving them separate IDs if they use the **inherited** query. This is also consistent with how pagination currently works for inherited queries. If I put 2 Query Loop blocks in a template that uses a Default query, both blocks are paginated simultaneously.

Thus, if a template contains multiple Query + Instant Search blocks, then the search should affect all of them equally.
In other words, you cannot perform independent searches in different Instant Seach blocks on a page that use an inherited query - they are always "in sync".

michalczaplinski committed:
Refactor how we pass the query_id and add a comment on the filter

michalczaplinski committed:
Set the search state in the interactivity state for inherited queries.

luisherranz commented:
> you cannot perform independent searches in different Instant Seach blocks on a page that use an inherited query - they are always "in sync"

That makes sense.

michalczaplinski committed:
Prevent default form submission on search submit.

michalczaplinski commented:
I've added support for multiple Default queries in [d1483a2](https://github.com/WordPress/gutenberg/pull/67289/commits/d1483a2016a26c617be6beceedf324e26ffc6e2e).

As commented above, the Default queries cannot run independently of one another - they are all "in sync".  I implemented it by using a setting the search value in the state if there is a Defalt (inherited) query and using a getter in the state: 

```php
// search/index.php
if ( $is_inherited ) {
    wp_interactivity_state( 'core/search', array( 'search' => $search ) );
}
```

```js
// search/view.js
get searchGetter() {
    const { isInherited, search } = getContext();
    return isInherited ? state.search : search;
},
```

Then in the `*updateSearch()` action, we either update the context (for Custom queries) or the state (for Default queries):

```js
// search/view.js
*updateSearch( e ) {
    //...
    const ctx = getContext();
    
    if ( ctx.isInherited ) {
	  state.search = value;
    } else {
	  ctx.search = value;
    }
    //...
}

michalczaplinski committed:
Add e2e tests for instant search with default query.

michalczaplinski committed:
Add an e2e test for multiple inherited and custom queries used in the same template

michalczaplinski committed:
PHP formatting

michalczaplinski committed:
Add a new request utils function to update or create a template.

We need this in case the `updateTemplate` function throws an error if the template is not found.

michalczaplinski commented:
I see an e2e test failure. 

I **think** it's spurious and this failing line should not even be there: https://github.com/WordPress/gutenberg/blob/78c032db15bdb7c71e735994facc49371e1aec44/test/e2e/specs/site-editor/template-registration.spec.js#L104 

There's no reason why the `Close` button should need to be clicked in that test 🤷 . Will have to investigate.

michalczaplinski commented:
The failing e2e test case should be fixed in https://github.com/WordPress/gutenberg/pull/68059

michalczaplinski committed:
Update e2e tests for Instant Search to verify block renaming in Inspector Controls and List View
